File: src/telegram/voice.ts
import { mkdir, writeFile } from "node:fs/promises";
import { join } from "node:path";
import type { Api, TelegramClient } from "telegram";
import { getPeerLabelFromMessage } from "@/telegram/utils";

const voicesDir = join(process.cwd(), "voices");

async function ensureVoicesDir(): Promise<void> {
    await mkdir(voicesDir, { recursive: true });
}

export function buildVoiceFileName(message: Api.Message): string {
    const peerId = getPeerLabelFromMessage(message);
    return `voice-${peerId}-${String(message.id)}-${Date.now()}.ogg`;
}

async function saveMediaFromMessage(client: TelegramClient, message: Api.Message, fileName: string): Promise<string> {
    await ensureVoicesDir();
    const destPath = join(voicesDir, fileName);
    const data = await client.downloadMedia(message, {});
    if (!data) throw new Error("Failed to download media");
    if (data instanceof Uint8Array) {
        await writeFile(destPath, Buffer.from(data));
    } else {
        throw new Error("Unexpected media type returned while downloading media");
    }
    return destPath;
}

export async function saveVoiceFromMessage(client: TelegramClient, message: Api.Message): Promise<string> {
    return saveMediaFromMessage(client, message, buildVoiceFileName(message));
}

export async function saveVideoNoteFromMessage(client: TelegramClient, message: Api.Message): Promise<string> {
    const peerId = getPeerLabelFromMessage(message);
    const fileName = `videonote-${peerId}-${String(message.id)}-${Date.now()}.mp4`;
    return saveMediaFromMessage(client, message, fileName);
}

File: src_py/telegram_utils/voice.py
import os
import time

from telethon import TelegramClient
from telethon.tl import types

from src_py.telegram_utils.utils import get_peer_label

VOICES_DIR = os.path.join(os.getcwd(), "voices")


async def _ensure_voices_dir() -> None:
    os.makedirs(VOICES_DIR, exist_ok=True)


def _build_voice_filename(message: types.Message) -> str:
    peer = get_peer_label(message)
    return f"voice-{peer}-{message.id}-{int(time.time() * 1000)}.ogg"


async def _save_media(
    client: TelegramClient, message: types.Message, filename: str
) -> str:
    await _ensure_voices_dir()
    dest = os.path.join(VOICES_DIR, filename)
    await client.download_media(message, file=dest)
    return dest


async def save_voice_from_message(
    client: TelegramClient, message: types.Message
) -> str:
    return await _save_media(client, message, _build_voice_filename(message))


async def save_video_note_from_message(
    client: TelegramClient, message: types.Message
) -> str:
    peer = get_peer_label(message)
    filename = f"videonote-{peer}-{message.id}-{int(time.time() * 1000)}.mp4"
    return await _save_media(client, message, filename)
